pkg/container/qemu_runner.go

```diff
@@ -0,0 +1,879 @@
+// as a microvm runner.
+func (bw *qemu) TestUsability(ctx context.Context) bool {
+	log := clog.FromContext(ctx)
+
+	arch := apko_types.Architecture(runtime.GOARCH)
+	if _, err := exec.LookPath(fmt.Sprintf("qemu-system-%s", arch.ToAPK())); err != nil {
+		log.Warnf("cannot use qemu for microvms: qemu-system-%s not found on $PATH", arch.ToAPK())
+		return false
+	}
+
+	return true
+}
+
+// OCIImageLoader used to load OCI images in, if needed. qemu does not need it.
+func (bw *qemu) OCIImageLoader() Loader {
+	return &qemuOCILoader{}
+}
+
+// TempDir returns the base for temporary directory. For qemu, this is empty.
+func (bw *qemu) TempDir() string {
+	return ""
+}
+
+// StartPod starts a pod if necessary.
+func (bw *qemu) StartPod(ctx context.Context, cfg *Config) error {
+	ctx, span := otel.Tracer("melange").Start(ctx, "qemu.StartPod")
+	defer span.End()
+
+	sshPort, err := randpomPortN()
+	if err != nil {
+		return err
+	}
+
+	cfg.SSHPort = strconv.Itoa(sshPort)
+
+	return createMicroVM(ctx, cfg)
+}
+
+// TerminatePod terminates a pod if necessary.  Not implemented
+// for Qemu runners.
+func (bw *qemu) TerminatePod(ctx context.Context, cfg *Config) error {
+	ctx, span := otel.Tracer("melange").Start(ctx, "qemu.TerminatePod")
+	defer span.End()
+
+	clog.FromContext(ctx).Info("qemu: sending shutdown signal")
+	err := sendSSHCommand(ctx,
+		"root",
+		"localhost",
+		cfg.SSHPort,
+		cfg,
+		nil,
+		nil,
+		nil,
+		nil,
+		[]string{"echo s > /proc/sysrq-trigger && echo o > /proc/sysrq-trigger&"},
+	)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// WorkspaceTar implements Runner
+func (bw *qemu) WorkspaceTar(ctx context.Context, cfg *Config) (io.ReadCloser, error) {
+	// default to root user but if a different user is specified
+	// we will use the embedded build:1000:1000 user
+	user := "root"
+	if cfg.RunAs != "" {
+		user = "build"
+	}
+
+	clog.FromContext(ctx).Infof("compressing remote workspace")
+	err := sendSSHCommand(ctx,
+		user,
+		"localhost",
+		cfg.SSHPort,
+		cfg,
+		nil,
+		nil,
+		nil,
+		nil,
+		[]string{"cd /home/build && tar cvzf melange-out.tar.gz melange-out"},
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	clog.FromContext(ctx).Infof("fetching remote workspace")
+	workspaceTar, err := os.OpenFile(filepath.Join(cfg.WorkspaceDir, "melange-out.tar.gz"), os.O_APPEND|os.O_WRONLY|os.O_CREATE, 0644)
+	if err != nil {
+		return nil, err
+	}
+	defer workspaceTar.Close()
+
+	// work around missing scp (needs openssh-sftp package), we just cat the file
+	// and pipe the output to our local file. It is potentially slower, but being
+	// a localhost interface, the performance penalty should be negligible.
+	err = sendSSHCommand(ctx,
+		user,
+		"localhost",
+		cfg.SSHPort,
+		cfg,
+		nil,
+		nil,
+		nil,
+		workspaceTar,
+		[]string{"cat /home/build/melange-out.tar.gz"},
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	return os.Open(filepath.Join(cfg.WorkspaceDir, "melange-out.tar.gz"))
```

Comment:
does the output have to be compressed?

it feels like compression and then decompression is going to be slower than skipping it on a virtio network.

Reply:
Not having proper scp and rsync I thought using a single connection to move a single file would be better
as throughput would be better than latencies in these type of connections (localhost)